src/main/java/software/amazon/smithy/lsp/ServerState.java

```diff
@@ -10,57 +10,71 @@
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.logging.Logger;
+import org.eclipse.lsp4j.FileEvent;
 import org.eclipse.lsp4j.WorkspaceFolder;
 import software.amazon.smithy.lsp.document.Document;
 import software.amazon.smithy.lsp.project.Project;
 import software.amazon.smithy.lsp.project.ProjectAndFile;
+import software.amazon.smithy.lsp.project.ProjectChange;
 import software.amazon.smithy.lsp.project.ProjectFile;
 import software.amazon.smithy.lsp.project.ProjectLoader;
 import software.amazon.smithy.lsp.protocol.LspAdapter;
 import software.amazon.smithy.lsp.util.Result;
 
 /**
  * Keeps track of the state of the server.
- *
- * @param detachedProjects Map of smithy file **uri** to detached project
- *                         for that file
- * @param attachedProjects Map of directory **path** to attached project roots
- * @param workspacePaths Paths to roots of each workspace open in the client
- * @param managedUris Uris of each file managed by the server/client, i.e.
- *                    files which may be updated by didChange
- * @param lifecycleManager Container for ongoing tasks
  */
-public record ServerState(
-        Map<String, Project> detachedProjects,
-        Map<String, Project> attachedProjects,
-        Set<Path> workspacePaths,
-        Set<String> managedUris,
-        DocumentLifecycleManager lifecycleManager
-) {
+public final class ServerState implements ManagedFiles {
     private static final Logger LOGGER = Logger.getLogger(ServerState.class.getName());
 
+    private final Map<String, Project> projects;
+    private final Set<Path> workspacePaths;
+    private final Set<String> managedUris;
+    private final FileTasks lifecycleTasks;
+
     /**
      * Create a new, empty server state.
      */
     public ServerState() {
-        this(
-                new HashMap<>(),
-                new HashMap<>(),
-                new HashSet<>(),
-                new HashSet<>(),
-                new DocumentLifecycleManager());
+        this.projects = new HashMap<>();
+        this.workspacePaths = new HashSet<>();
+        this.managedUris = new HashSet<>();
+        this.lifecycleTasks = new FileTasks();
     }
 
     /**
-     * @param uri Uri of the document to get
-     * @return The document if found and it is managed, otherwise {@code null}
+     * @return All projects tracked by the server.
      */
+    public Collection<Project> getAllProjects() {
+        return projects.values();
+    }
+
+    /**
+     * @return All files managed by the server, including their projects.
+     */
+    public Collection<ProjectAndFile> getAllManaged() {
```

Comment:
if this is public, you might want to add it to the ManagedFiles interface. just a thought

Reply:
Might make sense, or something similar. Right now ManagedFiles doesn't really know about Project, so we will have to see